sway/commands/bar/bind.c

```diff
@@ -9,10 +9,70 @@
 #include "log.h"
 #include "stringop.h"
 
-static struct cmd_results *bar_cmd_bind(int argc, char **argv, bool code) {
-	const char *command = code ? "bar bindcode" : "bar bindsym";
+static struct cmd_results *binding_add(struct bar_binding *binding,
+		list_t *mode_bindings) {
+	const char *name = get_mouse_button_name(binding->button);
+	bool overwritten = false;
+	for (int i = 0; i < mode_bindings->length; i++) {
+		struct bar_binding *other = mode_bindings->items[i];
+		if (other->button == binding->button &&
+				other->release == binding->release) {
+			overwritten = true;
+			mode_bindings->items[i] = binding;
+			free_bar_binding(other);
+			sway_log(SWAY_DEBUG, "[bar %s] Updated binding for %u (%s)%s",
+					config->current_bar->id, binding->button, name,
+					binding->release ? " - release" : "");
+			break;
+		}
+	}
+	if (!overwritten) {
+		list_add(mode_bindings, binding);
+		sway_log(SWAY_DEBUG, "[bar %s] Added binding for %u (%s)%s",
+				config->current_bar->id, binding->button, name,
+				binding->release ? " - release" : "");
+	}
+	return cmd_results_new(CMD_SUCCESS, NULL);
+}
+
+static struct cmd_results *binding_remove(struct bar_binding *binding,
+		list_t *mode_bindings) {
+	const char *name = get_mouse_button_name(binding->button);
+	for (int i = 0; i < mode_bindings->length; i++) {
+		struct bar_binding *other = mode_bindings->items[i];
+		if (other->button == binding->button &&
+				other->release == binding->release) {
+			free_bar_binding(other);
+			free_bar_binding(binding);
+			list_del(mode_bindings, i);
+			sway_log(SWAY_DEBUG, "[bar %s] Unbound binding for %u (%s)%s",
```

Comment:
This debug log needs to be moved before `binding` is freed. `binding->button` and `binding->release` will be use-after-frees with the current version.